python/pyspark/sql/dataframe.py

```diff
@@ -2536,6 +2536,28 @@ def withColumnRenamed(self, existing, new):
         """
         return DataFrame(self._jdf.withColumnRenamed(existing, new), self.sql_ctx)
 
+    def withMetadata(self, columnName, metadata):
+        """Returns a new :class:`DataFrame` by updating an existing column with metadata.
+
+        .. versionadded:: 3.3.0
+
+        Parameters
+        ----------
+        columnName : str
+            string, name of the existing column to update the metadata.
+        metadata : dict
+            dict, new metadata to be assigned to df.schema[columnName].metadata
+
+        Examples
+        --------
+        >>> df_meta = df.withMetadata('age', {'foo': 'bar'})
+        >>> df_meta.schema['age'].metadata
+        {'foo': 'bar'}
+        """
+        if not isinstance(metadata, dict):
```

Comment:
why restrict the type to be dict ? we should allow any types which can be converted to JSON.

Reply:
https://github.com/apache/spark/blob/master/python/pyspark/sql/column.py#L712 The existing API that takes the metadata also specified that the metadata should be a dict in the docstring. However, I'm also fine with not checking the dict type.

Reply:
@HyukjinKwon 

> https://github.com/apache/spark/blob/master/python/pyspark/sql/column.py#L712 The existing API that takes the metadata also specified that the metadata should be a dict in the docstring. However, I'm also fine with not checking the dict type.

But in code here https://github.com/apache/spark/blob/cabc36b54d7f6633d8b128e511e7049c475b919d/python/pyspark/sql/column.py#L747 it doesn't require metadata to be dict , so is it a doc error or code error there ?

Reply:
I think this is fine to check if it's an explicitly a dictionary. Following docstring sounds fine.